File: nixos/modules/system/etc/setup-etc.pl
use strict;
use File::Find;
use File::Copy;
use File::Path;
use File::Basename;
use File::Slurp;

my $etc = $ARGV[0] or die;
my $static = "/etc/static";

sub atomicSymlink {
    my ($source, $target) = @_;
    my $tmp = "$target.tmp";
    unlink $tmp;
    symlink $source, $tmp or return 0;
    if (rename $tmp, $target) {
        return 1;
    } else {
        unlink $tmp;
        return 0;
    }
}


# Atomically update /etc/static to point at the etc files of the
# current configuration.
atomicSymlink $etc, $static or die;

# Returns 1 if the argument points to the files in /etc/static.  That
# means either argument is a symlink to a file in /etc/static or a
# directory with all children being static.
sub isStatic {
    my $path = shift;

    if (-l $path) {
        my $target = readlink $path;
        return substr($target, 0, length "/etc/static/") eq "/etc/static/";
    }

    if (-d $path) {
        opendir DIR, "$path" or return 0;
        my @names = readdir DIR or die;
        closedir DIR;

        foreach my $name (@names) {
            next if $name eq "." || $name eq "..";
            unless (isStatic("$path/$name")) {
                return 0;
            }
        }
        return 1;
    }

    return 0;
}

# Use /etc/.clean to keep track of copied files and symlinks.
my @oldCopied = read_file("/etc/.clean", chomp => 1, err_mode => 'quiet');
open CLEAN, ">>/etc/.clean";


# For every file in the etc tree, create a corresponding symlink in
# /etc to /etc/static.  The indirection through /etc/static is to make
# switching to a new configuration somewhat more atomic.
my %created;
my @copied;

sub link {
    my $fn = substr $File::Find::name, length($etc) + 1 or next;

    # nixos-enter sets up /etc/resolv.conf as a bind mount, so skip it.
    if ($fn eq "resolv.conf" and $ENV{'IN_NIXOS_ENTER'}) {
        return;
    }

    my $target = "/etc/$fn";
    File::Path::make_path(dirname $target);
    $created{$fn} = 1;

    # Rename doesn't work if target is directory.
    if (-l $_ && -d $target) {
        if (isStatic $target) {
            rmtree $target or warn;
        } else {
            warn "$target directory contains user files. Symlinking may fail.";
        }
    }

    if (-e "$_.mode") {
        my $mode = read_file("$_.mode"); chomp $mode;
        if ($mode eq "direct-symlink") {
            atomicSymlink readlink("$static/$fn"), $target or warn "could not create symlink $target";
        } else {
            my $uid = read_file("$_.uid"); chomp $uid;
            my $gid = read_file("$_.gid"); chomp $gid;
            copy "$static/$fn", "$target.tmp" or warn;
            $uid = getpwnam $uid unless $uid =~ /^\+/;
            $gid = getgrnam $gid unless $gid =~ /^\+/;
            chown int($uid), int($gid), "$target.tmp" or warn;
            chmod oct($mode), "$target.tmp" or warn;
            unless (rename "$target.tmp", $target) {
                warn "could not create target $target";
                unlink "$target.tmp";
            }
        }
        push @copied, $fn;
        print CLEAN "$fn\n";
    } elsif (-l "$_") {
        atomicSymlink "$static/$fn", $target or warn "could not create symlink $target";
        push @copied, $fn;
        print CLEAN "$fn\n";
    }
}

find(\&link, $etc);


# Delete files and symlinks that were copied in a previous version but not in the
# current.
foreach my $fn (@oldCopied) {
    if (!defined $created{$fn}) {
        $fn = "/etc/$fn";
        print STDERR "removing obsolete file ‘$fn’...\n";
        unlink "$fn";
    }
}


# Rewrite /etc/.clean.
close CLEAN;
write_file("/etc/.clean", map { "$_\n" } sort @copied);

# Create /etc/NIXOS tag if not exists.
# When /etc is not on a persistent filesystem, it will be wiped after reboot,
# so we need to check and re-create it during activation.
open TAG, ">>/etc/NIXOS";
close TAG;

File: nixos/modules/system/etc/setup-etc.py
import argparse
import os
import os.path
from pathlib import Path
from pwd import getpwnam
from grp import getgrnam
import shutil
import sys
from typing import List, Tuple, Set

STATIC = Path("/etc/static")
CLEAN_FILE = Path("/etc/.clean")

def warn(msg: str) -> None:
    print('warning:', msg, file=sys.stderr)

def atomic_symlink(source: Path, target: Path) -> None:
    tmp = Path("{target}.tmp".format(target=target))
    tmp.unlink()
    tmp.symlink_to(source)
    tmp.rename(target)

def atomic_symlink_or_warn(source: Path, target: Path) -> None:
    try:
        atomic_symlink(source, target)
    except IOError as exc:
        warn("error creating link '{target}': {exc}". \
             format(target=target, exc=exc))

def is_static(path: Path) -> bool:
    """
    Returns True if the argument points to the files in /etc/static. That means
    either argument is a symlink to a file in /etc/static or a directory with
    all children being static.
    """

    if path.is_symlink():
        target = path.resolve()
        res = os.path.commonprefix([target, STATIC]) == STATIC  # type: bool
        return res

    if path.is_dir():
        return all(is_static(ent) for ent in path.iterdir())

    return False

def cleanup() -> None:
    """
    Remove dangling symlinks that point to /etc/static.  These are
    configuration files that existed in a previous configuration but not
    in the current one.  For efficiency, don't look under /etc/nixos
    (where all the NixOS sources live).
    """
    for root, dirs, files in os.walk('/etc'):
        if root == '/etc':
            dirs.remove('nixos')

        for f in files:
            path = Path(root) / Path(f)
            if path.is_symlink():
                target = path.resolve()
                if os.path.commonprefix([target, STATIC]) == STATIC:
                    x = STATIC / path.relative_to('/etc')
                    if x.is_symlink():
                        print("removing obsolete symlink '{path}'".format(path=path),
                              file=sys.stderr)
                        path.unlink()

def link_statics(etc: Path) -> Tuple[Set[Path], Set[Path]]:
    """
    For every file in the etc tree, create a corresponding symlink in
    /etc to /etc/static.  The indirection through /etc/static is to make
    switching to a new configuration somewhat more atomic.
    """

    clean = open(CLEAN_FILE, 'a')
    created = set()
    copied = set()
    for root, dirs, files in os.walk(etc):
        for f in files:
            path = Path(root) / Path(f)
            if os.path.commonprefix([etc, path]) != etc:
                continue
            fn = path.relative_to(etc)
            target = Path("/etc") / f
            target.parent.mkdir(parents=True)
            created.add(fn)

            # Rename doesn't work if target is directory.
            if Path(f).is_symlink() and target.is_dir():
                if is_static(target):
                    try:
                        shutil.rmtree(target)
                    except IOError as exc:
                        warn("error trying to remove '{target}': {exc}". \
                             format(target=target, exc=exc))
                else:
                    warn("$target directory contains user files. Symlinking may fail.")

            meta_f = Path("{f}.mode".format(f=f))
            if meta_f.exists():
                mode = open(meta_f, 'r').read().strip()
                if mode == 'direct-symlink':
                    src = (STATIC / fn).resolve()
                    atomic_symlink_or_warn(src, target)
                else:
                    user = open("{f}.uid".format(f=f)).read().strip()
                    grp = open("{f}.gid".format(f=f)).read().strip()
                    uid = getpwnam(user).pw_uid
                    gid = getgrnam(grp).gr_gid

                    src = STATIC / fn
                    tmp = Path("{target}.tmp".format(target=target))
                    try:
                        shutil.copyfile(src, tmp)
                    except IOError as exc:
                        warn("error copying from '{src}' to '{tmp}': {exc}". \
                             format(src=src, tmp=tmp, exc=exc))

                    try:
                        os.chown(tmp, uid, gid)
                    except IOError as exc:
                        warn("error setting owner of '{tmp}': {exc}". \
                             format(tmp=tmp, exc=exc))

                    try:
                        os.chmod(tmp, int(mode, 8))
                    except IOError as exc:
                        warn("error setting mode of '{tmp}': {exc}". \
                             format(tmp=tmp, exc=exc))

                    try:
                        tmp.rename(target)
                    except IOError as exc:
                        warn("error renaming '{tmp}' to '{target}': {exc}". \
                             format(tmp=tmp, target=target, exc=exc))

                copied.add(fn)
                print(fn, file=clean)

            elif Path(f).is_symlink():
                atomic_symlink_or_warn(STATIC / fn, target)

    return (created, copied)

def main() -> None:
    parser = argparse.ArgumentParser()
    parser.add_argument('etc')
    args = parser.parse_args()
    etc = args.etc

    # Atomically update /etc/static to point at the etc files of the
    # current configuration.
    atomic_symlink(etc, STATIC)

    cleanup()

    # Use /etc/.clean to keep track of copied files.
    old_copied = []  # type: List[str]
    if CLEAN_FILE.is_file():
        old_copied += [ line.strip() for line in CLEAN_FILE.read_text().split('\n') ]

    created, copied = link_statics(etc)

    # Delete files that were copied in a previous version but not in the
    # current.
    for f in old_copied:
        if f not in created:
            fn = Path("/etc") / f
            print("removing obsolete file '{f}'...".format(f=f),
                  file=sys.stderr)
            fn.unlink()

    # Rewrite /etc/.clean
    CLEAN_FILE.write_text('\n'.join(str(p) for p in copied))

if __name__ == "__main__":
    main()
